data/tokenizer.py

```diff
@@ -0,0 +1,238 @@
+            range(self.quantization_cfg.duration),
+        )
+
+        for pitch in range(21, 109):
+            self.vocab.append(f"{pitch}p")
+        for start, duration in time_tokens_product:
+            self.vocab.append(f"{start:0.0f}-{duration:0.0f}")
+        # in hf T5 and BART we will use common vocabulary for tgt and src tokens ¯\_(ツ)_/¯
+        # luckily, MultiTok is ideal for this
+        # (UNLESS we will manually swap a final layer (somehow))
+        # TODO: find out how sharing tokens effects model performance. (1v, 2v, 3v for 3 bins and also for 128 bins)
+        for velocity in range(128):
+            self.vocab.append(f"{velocity:0.0f}v")
+
+    def tokenize_src(self, record: dict) -> list[str]:
+        tokens = []
+        n_samples = len(record[self.keys[0]])
+        for idx in range(n_samples):
+            # append p and v so that there are no duplicate tokens
+            pitch_token = f"{record[self.keys[0]][idx]:0.0f}p"
+            time_token = f"{record[self.keys[1]][idx]}-{record[self.keys[2]][idx]:0.0f}"
+            velocity_token = f"{record[self.keys[3]][idx]:0.0f}v"
+            tokens += [pitch_token, time_token, velocity_token]
+
+        return tokens
+
+    def tokenize_tgt(self, record: dict) -> list[str]:
+        tokens = [f"{velocity:0.0f}v" for velocity in record["velocity"]]
+        return tokens
+
+    def untokenize_src(self, tokens: list[str]) -> pd.DataFrame:
+        samples = []
+        buff = []
+        if tokens[0] == "<CLS>":
+            # get rid of cls token
+            tokens = tokens[1:]
+
+        for idx, token in enumerate(tokens):
+            if token == "<pad>":
+                break
+            if idx % 3 == 0:
+                # pitch token
+                buff.clear()
+                # [:-1] to clear p letter from a token
+                buff.append(eval(token[:-1]))
+            elif idx % 3 == 1:
+                # time token
+                buff.append(eval(txt) for txt in token.split("-"))
+            else:
+                # velocity token
+                # [:-1] to clear v letter from the token
+                buff.append(eval(token[:-1]))
+                samples.append(buff)
+
+        df = pd.DataFrame(samples, columns=self.keys)
+
+        return df
+
+    def untokenize_tgt(self, tokens: list[str]) -> list[int]:
+        # [:-1] to remove v letter from the end of a token
+        velocities = [int(token[:-1]) for token in tokens if token not in self.specials]
+
+        return velocities
+
+
+class MultiStartEncoder(MultiTokEncoder):
+    def __init__(self, quantization_cfg: DictConfig, keys: list[str] = None, tgt_bins: int = 650):
+        super().__init__()
+        self.quantization_cfg = quantization_cfg
+        self.tgt_bins = tgt_bins
+        if keys is None:
+            self.keys = ["pitch", "dstart_bin", "duration_bin", "velocity_bin"]
+        else:
+            self.keys = keys
+
+        self.specials = ["<CLS>", "<pad>"]
```

Comment:
```suggestion
        self.specials = ["<CLS>", "<PAD>"]
```

Unless there is a reason for this lack of lower/upper case consistency :)